File: acp_ui_flutter/lib/core/router/smart_router.dart
/// Smart Router - 任务复杂度分析与路由决策
///
/// 调用 Tauri 后端的 analyze_task_complexity 命令

import 'dart:async';
import 'dart:convert';
import 'package:flutter/foundation.dart';

/// 任务复杂度级别
enum TaskComplexity {
  simple,
  medium,
  complex,
  veryComplex,
}

/// 任务类型
enum TaskType {
  frontend,
  backend,
  fullstack,
  devops,
  testing,
  documentation,
  refactoring,
  bugfix,
}

/// 输入类型
enum InputType {
  text,
  image,
  document,
  code,
  command,
}

/// 路由目标
enum RouteTarget {
  claudeCodeHaiku,
  claudeCodeSonnet,
  codexHaiku,
  codexSonnet,
  team,
  humanReview,
}

/// 评估方法
enum EvaluationMethod {
  heuristic,
  structural,
  llmAssisted,
}

/// 路由决策结果
class RouteDecision {
  final String id;
  final TaskType taskType;
  final InputType inputType;
  final TaskComplexity complexity;
  final RouteTarget routeTarget;
  final String reason;
  final int agentLoad;
  final double historicalSuccessRate;
  final EvaluationMethod evaluationMethod;
  final DateTime createdAt;

  RouteDecision({
    required this.id,
    required this.taskType,
    required this.inputType,
    required this.complexity,
    required this.routeTarget,
    required this.reason,
    this.agentLoad = 0,
    this.historicalSuccessRate = 0.8,
    required this.evaluationMethod,
    required this.createdAt,
  });

  factory RouteDecision.fromJson(Map<String, dynamic> json) {
    return RouteDecision(
      id: json['id'] as String,
      taskType: _parseTaskType(json['task_type'] as String),
      inputType: _parseInputType(json['input_type'] as String),
      complexity: _parseComplexity(json['complexity'] as String),
      routeTarget: _parseRouteTarget(json['route_target'] as String),
      reason: json['reason'] as String,
      agentLoad: json['agent_load'] as int? ?? 0,
      historicalSuccessRate: (json['historical_success_rate'] as num?)?.toDouble() ?? 0.8,
      evaluationMethod: _parseEvaluationMethod(json['evaluation_method'] as String),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  static TaskType _parseTaskType(String s) {
    return TaskType.values.firstWhere(
      (e) => e.name == s,
      orElse: () => TaskType.fullstack,
    );
  }

  static InputType _parseInputType(String s) {
    return InputType.values.firstWhere(
      (e) => e.name == s,
      orElse: () => InputType.text,
    );
  }

  static TaskComplexity _parseComplexity(String s) {
    return TaskComplexity.values.firstWhere(
      (e) => e.name.replaceAll('_', '') == s.replaceAll('_', ''),
      orElse: () => TaskComplexity.medium,
    );
  }

  static RouteTarget _parseRouteTarget(String s) {
    return RouteTarget.values.firstWhere(
      (e) => e.name.replaceAll('_', '') == s.replaceAll('-', ''),
      orElse: () => RouteTarget.team,
    );
  }

  static EvaluationMethod _parseEvaluationMethod(String s) {
    return EvaluationMethod.values.firstWhere(
      (e) => e.name.replaceAll('_', '') == s.replaceAll('_', ''),
      orElse: () => EvaluationMethod.heuristic,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'task_type': taskType.name,
    'input_type': inputType.name,
    'complexity': complexity.name,
    'route_target': routeTarget.name.replaceAll('_', '-'),
    'reason': reason,
    'agent_load': agentLoad,
    'historical_success_rate': historicalSuccessRate,
    'evaluation_method': evaluationMethod.name,
    'created_at': createdAt.toIso8601String(),
  };
}

/// Smart Router 服务
class SmartRouter {
  final WebSocketService? _wsService;

  SmartRouter({WebSocketService? wsService}) : _wsService = wsService;

  /// 分析任务复杂度
  Future<RouteDecision> analyze(
    String input,
    InputType inputType,
  ) async {
    if (_wsService != null && _wsService!.isConnectedGetter) {
      // WebSocket 方式调用 Tauri 命令
      final request = {
        'command': 'analyze_task_complexity',
        'args': {
          'input': input,
          'input_type': inputType.name,
        },
      };

      // TODO: 实现 WebSocket 响应等待
      // final response = await _wsService.sendAndWait(request);
      // return RouteDecision.fromJson(response);
    }

    // 本地启发式分析（备用）
    return _localAnalyze(input, inputType);
  }

  /// 本地启发式分析
  RouteDecision _localAnalyze(String input, InputType inputType) {
    final complexity = _determineComplexity(input);
    final taskType = _determineTaskType(input);
    final target = _selectTarget(taskType, complexity);

    return RouteDecision(
      id: 'local-${DateTime.now().millisecondsSinceEpoch}',
      taskType: taskType,
      inputType: inputType,
      complexity: complexity,
      routeTarget: target,
      reason: 'Local heuristic analysis',
      evaluationMethod: EvaluationMethod.heuristic,
      createdAt: DateTime.now(),
    );
  }

  TaskComplexity _determineComplexity(String input) {
    final lower = input.toLowerCase();
    if (lower.contains('修改一行') || lower.contains('fix') || lower.contains('bug')) {
      return TaskComplexity.simple;
    }
    if (lower.contains('refactor') || lower.contains('重构') || lower.contains('添加')) {
      return TaskComplexity.medium;
    }
    if (lower.contains('跨模块') || lower.contains('新功能') || lower.contains('feature')) {
      return TaskComplexity.complex;
    }
    if (lower.contains('多agent') || lower.contains('system') || lower.contains('架构')) {
      return TaskComplexity.veryComplex;
    }
    return TaskComplexity.medium;
  }

  TaskType _determineTaskType(String input) {
    final lower = input.toLowerCase();
    if (lower.contains('vue') || lower.contains('ui') || lower.contains('前端')) {
      return TaskType.frontend;
    }
    if (lower.contains('api') || lower.contains('backend') || lower.contains('数据库')) {
      return TaskType.backend;
    }
    if (lower.contains('test') || lower.contains('测试')) {
      return TaskType.testing;
    }
    if (lower.contains('doc') || lower.contains('文档')) {
      return TaskType.documentation;
    }
    return TaskType.fullstack;
  }

  RouteTarget _selectTarget(TaskType taskType, TaskComplexity complexity) {
    if (complexity == TaskComplexity.veryComplex) {
      return RouteTarget.team;
    }
    switch (taskType) {
      case TaskType.frontend:
        return complexity == TaskComplexity.simple
            ? RouteTarget.claudeCodeHaiku
            : RouteTarget.claudeCodeSonnet;
      case TaskType.backend:
        return complexity == TaskComplexity.simple
            ? RouteTarget.codexHaiku
            : RouteTarget.codexSonnet;
      default:
        return RouteTarget.claudeCodeSonnet;
    }
  }
}

/// WebSocketService 类型占位（实际从 services 导入）
class WebSocketService {
  final String url;
  WebSocketService({required this.url});
  bool isConnectedGetter = false;
}

File: src-tauri/src/smart_router.rs
//! Smart Router - Three-layer progressive complexity evaluation
//!
//! Implements intelligent routing: Heuristic → Structural → LLM-assisted

use serde::{Deserialize, Serialize};
use chrono::{DateTime, Utc};
use std::collections::HashMap;

/// Task complexity level
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum TaskComplexity {
    Simple,     // Single file, minor changes
    Medium,     // Multiple files, refactoring
    Complex,    // Cross-module, new features
    VeryComplex, // Multi-agent needed
}

#[allow(dead_code)]
impl TaskComplexity {
    pub fn as_str(&self) -> &'static str {
        match self {
            TaskComplexity::Simple => "simple",
            TaskComplexity::Medium => "medium",
            TaskComplexity::Complex => "complex",
            TaskComplexity::VeryComplex => "very_complex",
        }
    }
}

/// Task type classification
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TaskType {
    Frontend,       // UI, Vue components
    Backend,        // API, database
    Fullstack,      // Both frontend and backend
    DevOps,         // CI/CD, deployment
    Testing,        // Tests, QA
    Documentation,  // Docs, comments
    Refactoring,    // Code cleanup
    BugFix,         // Fix issues
}

#[allow(dead_code)]
impl TaskType {
    pub fn as_str(&self) -> &'static str {
        match self {
            TaskType::Frontend => "frontend",
            TaskType::Backend => "backend",
            TaskType::Fullstack => "fullstack",
            TaskType::DevOps => "devops",
            TaskType::Testing => "testing",
            TaskType::Documentation => "documentation",
            TaskType::Refactoring => "refactoring",
            TaskType::BugFix => "bugfix",
        }
    }

    pub fn from_str(s: &str) -> Result<Self, String> {
        match s {
            "frontend" => Ok(TaskType::Frontend),
            "backend" => Ok(TaskType::Backend),
            "fullstack" => Ok(TaskType::Fullstack),
            "devops" => Ok(TaskType::DevOps),
            "testing" => Ok(TaskType::Testing),
            "documentation" => Ok(TaskType::Documentation),
            "refactoring" => Ok(TaskType::Refactoring),
            "bugfix" => Ok(TaskType::BugFix),
            other => Err(format!("Unknown task type: {}", other)),
        }
    }
}

/// Input type classification
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum InputType {
    Text,       // Plain text request
    Image,      // Image input
    Document,   // Document/file input
    Code,       // Code snippet
    Command,    // Direct command
}

#[allow(dead_code)]
impl InputType {
    pub fn as_str(&self) -> &'static str {
        match self {
            InputType::Text => "text",
            InputType::Image => "image",
            InputType::Document => "document",
            InputType::Code => "code",
            InputType::Command => "command",
        }
    }
}

/// Route target - which agent/team to use
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum RouteTarget {
    ClaudeCodeHaiku,    // Simple tasks, fast response
    ClaudeCodeSonnet,   // Medium tasks, balanced
    CodexHaiku,         // Backend simple
    CodexSonnet,        // Backend complex
    Team,               // Multi-agent collaboration
    HumanReview,        // Needs human decision
}

#[allow(dead_code)]
impl RouteTarget {
    pub fn as_str(&self) -> &'static str {
        match self {
            RouteTarget::ClaudeCodeHaiku => "claude-code-haiku",
            RouteTarget::ClaudeCodeSonnet => "claude-code-sonnet",
            RouteTarget::CodexHaiku => "codex-haiku",
            RouteTarget::CodexSonnet => "codex-sonnet",
            RouteTarget::Team => "team",
            RouteTarget::HumanReview => "human-review",
        }
    }
}

/// Route decision record
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RouteDecision {
    pub id: String,
    pub input_id: String,
    pub task_type: TaskType,
    pub input_type: InputType,
    pub complexity: TaskComplexity,
    pub route_target: RouteTarget,
    pub reason: String,
    pub agent_load: u32,
    pub historical_success_rate: f64,
    pub evaluation_method: EvaluationMethod,
    pub created_at: DateTime<Utc>,
}

/// Evaluation method used
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum EvaluationMethod {
    Heuristic,   // Zero-cost rule matching
    Structural,  // Low-cost AST analysis
    LLMAssisted, // High-cost LLM evaluation
}

#[allow(dead_code)]
impl EvaluationMethod {
    pub fn as_str(&self) -> &'static str {
        match self {
            EvaluationMethod::Heuristic => "heuristic",
            EvaluationMethod::Structural => "structural",
            EvaluationMethod::LLMAssisted => "llm_assisted",
        }
    }
}

/// Task analyzer for complexity evaluation
pub struct TaskAnalyzer {
    heuristic_rules: HashMap<String, TaskComplexity>,
    structural_patterns: HashMap<String, TaskType>,
}

impl TaskAnalyzer {
    pub fn new() -> Self {
        let mut heuristic_rules = HashMap::new();

        // Simple task patterns
        heuristic_rules.insert("修改一行".to_string(), TaskComplexity::Simple);
        heuristic_rules.insert("修复bug".to_string(), TaskComplexity::Simple);
        heuristic_rules.insert("添加注释".to_string(), TaskComplexity::Simple);
        heuristic_rules.insert("格式化代码".to_string(), TaskComplexity::Simple);

        // Medium task patterns
        heuristic_rules.insert("重构".to_string(), TaskComplexity::Medium);
        heuristic_rules.insert("添加功能".to_string(), TaskComplexity::Medium);
        heuristic_rules.insert("优化性能".to_string(), TaskComplexity::Medium);
        heuristic_rules.insert("添加测试".to_string(), TaskComplexity::Medium);

        // Complex task patterns
        heuristic_rules.insert("跨模块".to_string(), TaskComplexity::Complex);
        heuristic_rules.insert("新功能".to_string(), TaskComplexity::Complex);
        heuristic_rules.insert("架构调整".to_string(), TaskComplexity::Complex);
        heuristic_rules.insert("集成".to_string(), TaskComplexity::Complex);

        // Very complex patterns
        heuristic_rules.insert("多agent".to_string(), TaskComplexity::VeryComplex);
        heuristic_rules.insert("大型重构".to_string(), TaskComplexity::VeryComplex);
        heuristic_rules.insert("系统设计".to_string(), TaskComplexity::VeryComplex);

        let mut structural_patterns = HashMap::new();
        structural_patterns.insert("vue".to_string(), TaskType::Frontend);
        structural_patterns.insert("component".to_string(), TaskType::Frontend);
        structural_patterns.insert("api".to_string(), TaskType::Backend);
        structural_patterns.insert("database".to_string(), TaskType::Backend);
        structural_patterns.insert("test".to_string(), TaskType::Testing);
        structural_patterns.insert("doc".to_string(), TaskType::Documentation);

        Self {
            heuristic_rules,
            structural_patterns,
        }
    }

    /// Analyze task using three-layer evaluation
    pub fn analyze(&self, input: &str, input_type: InputType) -> RouteDecision {
        let id = uuid::Uuid::new_v4().to_string();

        // Layer 1: Heuristic evaluation (zero cost)
        let (complexity, task_type) = self.heuristic_eval(input);

        // If heuristic is confident, use it (80% of requests)
        if complexity != TaskComplexity::Complex {
            let target = self.select_target(&task_type, &complexity, &input_type, 0, 0.8);

            return RouteDecision {
                id,
                input_id: String::new(),
                task_type,
                input_type,
                complexity,
                route_target: target,
                reason: "Heuristic match".to_string(),
                agent_load: 0,
                historical_success_rate: 0.8,
                evaluation_method: EvaluationMethod::Heuristic,
                created_at: Utc::now(),
            };
        }

        // Layer 2: Structural evaluation (low cost)
        let structural_result = self.structural_eval(input);
        let (complexity2, task_type2) = structural_result;

        // If structural gives clear answer, use it (15% of requests)
        if complexity2 != TaskComplexity::Complex {
            let target = self.select_target(&task_type2, &complexity2, &input_type, 0, 0.85);

            return RouteDecision {
                id,
                input_id: String::new(),
                task_type: task_type2,
                input_type,
                complexity: complexity2,
                route_target: target,
                reason: "Structural analysis".to_string(),
                agent_load: 0,
                historical_success_rate: 0.85,
                evaluation_method: EvaluationMethod::Structural,
                created_at: Utc::now(),
            };
        }

        // Layer 3: LLM-assisted evaluation (high cost, 5% of requests)
        // In production, this would call an LLM to analyze
        let target = RouteTarget::Team; // Default to team for complex

        RouteDecision {
            id,
            input_id: String::new(),
            task_type: TaskType::Fullstack,
            input_type,
            complexity: TaskComplexity::Complex,
            route_target: target,
            reason: "LLM-assisted evaluation recommended Team".to_string(),
            agent_load: 0,
            historical_success_rate: 0.9,
            evaluation_method: EvaluationMethod::LLMAssisted,
            created_at: Utc::now(),
        }
    }

    /// Heuristic evaluation - rule matching
    fn heuristic_eval(&self, input: &str) -> (TaskComplexity, TaskType) {
        let input_lower = input.to_lowercase();

        // Check complexity patterns
        for (pattern, complexity) in &self.heuristic_rules {
            if input_lower.contains(pattern) {
                // Determine task type from keywords
                let task_type = self.determine_task_type(&input_lower);
                return (*complexity, task_type);
            }
        }

        // Default to medium complexity
        (TaskComplexity::Medium, TaskType::Fullstack)
    }

    /// Structural evaluation - keyword counting
    fn structural_eval(&self, input: &str) -> (TaskComplexity, TaskType) {
        let input_lower = input.to_lowercase();

        // Count file references
        let file_count = input_lower.matches("文件").count()
            + input_lower.matches(".vue").count()
            + input_lower.matches(".ts").count()
            + input_lower.matches(".rs").count();

        // Count module references
        let module_count = input_lower.matches("模块").count()
            + input_lower.matches("组件").count()
            + input_lower.matches("服务").count();

        // Determine complexity by counts
        let complexity = if file_count <= 1 && module_count == 0 {
            TaskComplexity::Simple
        } else if file_count <= 3 && module_count <= 1 {
            TaskComplexity::Medium
        } else if file_count <= 5 && module_count <= 3 {
            TaskComplexity::Complex
        } else {
            TaskComplexity::VeryComplex
        };

        let task_type = self.determine_task_type(&input_lower);

        (complexity, task_type)
    }

    /// Determine task type from keywords
    fn determine_task_type(&self, input: &str) -> TaskType {
        for (pattern, task_type) in &self.structural_patterns {
            if input.contains(pattern) {
                return task_type.clone();
            }
        }

        if input.contains("前端") || input.contains("ui") || input.contains("界面") {
            TaskType::Frontend
        } else if input.contains("后端") || input.contains("api") || input.contains("数据库") {
            TaskType::Backend
        } else if input.contains("测试") || input.contains("qa") {
            TaskType::Testing
        } else {
            TaskType::Fullstack
        }
    }

    /// Select route target based on analysis
    fn select_target(
        &self,
        task_type: &TaskType,
        complexity: &TaskComplexity,
        input_type: &InputType,
        _agent_load: u32,
        _success_rate: f64,
    ) -> RouteTarget {
        // Image input prefers Claude Code (visual capability)
        if *input_type == InputType::Image {
            return RouteTarget::ClaudeCodeSonnet;
        }

        // Document input prefers Team (分工处理)
        if *input_type == InputType::Document {
            return RouteTarget::Team;
        }

        // Route by task type and complexity
        match (task_type, complexity) {
            (TaskType::Frontend, TaskComplexity::Simple) => RouteTarget::ClaudeCodeHaiku,
            (TaskType::Frontend, TaskComplexity::Medium) => RouteTarget::ClaudeCodeSonnet,
            (TaskType::Frontend, TaskComplexity::Complex) => RouteTarget::Team,
            (TaskType::Backend, TaskComplexity::Simple) => RouteTarget::CodexHaiku,
            (TaskType::Backend, TaskComplexity::Medium) => RouteTarget::CodexSonnet,
            (TaskType::Backend, TaskComplexity::Complex) => RouteTarget::Team,
            (TaskType::Fullstack, TaskComplexity::Simple) => RouteTarget::ClaudeCodeHaiku,
            (TaskType::Fullstack, TaskComplexity::Medium) => RouteTarget::ClaudeCodeSonnet,
            (TaskType::Fullstack, TaskComplexity::Complex) => RouteTarget::Team,
            (TaskType::Testing, _) => RouteTarget::ClaudeCodeSonnet,
            (TaskType::Documentation, _) => RouteTarget::ClaudeCodeHaiku,
            (TaskType::Refactoring, TaskComplexity::Simple) => RouteTarget::ClaudeCodeHaiku,
            (TaskType::Refactoring, TaskComplexity::Medium) => RouteTarget::ClaudeCodeSonnet,
            (TaskType::Refactoring, TaskComplexity::Complex) => RouteTarget::Team,
            (TaskType::BugFix, _) => RouteTarget::ClaudeCodeSonnet,
            (TaskType::DevOps, _) => RouteTarget::CodexSonnet,
            (_, TaskComplexity::VeryComplex) => RouteTarget::Team,
        }
    }
}

impl Default for TaskAnalyzer {
    fn default() -> Self {
        Self::new()
    }
}